File: euler/tyama_euler97.py
#!/usr/bin/python
M=1000000000000
r=0
for _ in range(int(raw_input())):
	a,b,c,d=map(int,raw_input().split())
	r=(r+a*pow(b,c,M)+d)%M
print('%012d'%r)

File: euler/tyama_euler97.rb
#!/usr/bin/ruby
X = 12 # 10
MOD = 10**X
class Integer
	def pow_binary_mod(y)
		x = self
		z = 1
		while y!=0
			z=z*x%MOD if y&1!=0
			x=x*x%MOD
			y>>=1
		end
		z
	end
end
r=0
gets.to_i.times{
	a,b,c,d=gets.split.map(&:to_i)
	r=(r+a*b.pow_binary_mod(c)+d)%MOD
}
puts '%0*d'%[X,r]
__END__
1
28433 2 7830457 1